xrtl/gfx/es3/es3_shader.cc

```diff
@@ -0,0 +1,199 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xrtl/gfx/es3/es3_shader.h"
+
+#include <spirv_glsl.hpp>
+
+#include <utility>
+#include <vector>
+
+#include "xrtl/base/debugging.h"
+#include "xrtl/base/tracing.h"
+
+namespace xrtl {
+namespace gfx {
+namespace es3 {
+
+ES3Shader::ES3Shader(ref_ptr<ES3PlatformContext> platform_context,
+                     std::string entry_point)
+    : platform_context_(platform_context),
+      entry_point_(std::move(entry_point)) {}
+
+ES3Shader::~ES3Shader() {
+  ES3PlatformContext::ThreadLock context_lock(
+      ES3PlatformContext::AcquireThreadContext(platform_context_));
+  if (shader_id_) {
+    glDeleteShader(shader_id_);
+  }
+}
+
+bool ES3Shader::CompileSource(GLenum shader_type,
+                              ArrayView<const char*> sources) {
+  std::vector<StringView> string_view_sources;
+  string_view_sources.reserve(sources.size());
+  for (const auto& source : sources) {
+    string_view_sources.push_back(StringView(source));
+  }
+  return CompileSource(shader_type, string_view_sources);
+}
+
+bool ES3Shader::CompileSource(GLenum shader_type,
+                              ArrayView<std::string> sources) {
+  std::vector<StringView> string_view_sources;
+  string_view_sources.reserve(sources.size());
+  for (const auto& source : sources) {
+    string_view_sources.push_back(StringView(source));
+  }
+  return CompileSource(shader_type, string_view_sources);
+}
+
+bool ES3Shader::CompileSource(GLenum shader_type,
+                              ArrayView<StringView> sources) {
+  WTF_SCOPE0("ES3Shader#CompileSource");
+  ES3PlatformContext::ThreadLock context_lock(
+      ES3PlatformContext::AcquireThreadContext(platform_context_));
+
+  shader_type_ = shader_type;
+  shader_id_ = glCreateShader(shader_type_);
+
+  {
+    // MESA has leaks in its shader compiler so we disable temporarily.
+    debugging::LeakCheckDisabler leak_check_disabler;
+
+    // Attach source now and start compilation.
+    // On browsers without async compilation this will block the thread.
```

Comment:
nit: "On systems" instead of "On browsers" (or could still mention WebGL / browsers if you want)